### Task 2: Dataset Creation 
1. Create student result dataset 
2. Create loan approval dataset 
3. Create disease prediction dataset 

### Task 3: Model Building 
1. Create Decision Tree model 
2. Train model 
3. Make predictions 

### Task 4: Prediction Practice 
Predict: 
1. Student result 
2. Disease status 
3. Loan approval 

In [16]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier, export_text
import warnings
warnings.filterwarnings('ignore')

#### 1. student result dataset

In [17]:
data = {
    "Hours_Studied":[2,3,4,5,6,7,8,9,4,7],
    "Attendance":[55,60,65,70,75,80,85,90,72,88],
    "Assignments_Submitted":[3,4,5,6,7,8,9,10,6,9],
    "Result":[0,0,0,0,1,1,1,1,0,1]}
student_df = pd.DataFrame(data) 


In [18]:
X_student_df = student_df[["Hours_Studied","Attendance","Assignments_Submitted"]]
y_student_df= student_df["Result"] 
#create model
model_s = DecisionTreeClassifier(max_depth=3, random_state=42)
#train model
model_s.fit(X_student_df,y_student_df)
#prediction
p_1 = model_s.predict([[6.5,80,7]])
print("p1:",p_1)

p1: [1]


#### 2. loan approval dataset

In [19]:
#loan approval dataset
data = {'Income': [25000,30000,35000,45000,50000,60000,70000,80000,90000,100000],
    'CreditScore': [500,550,600,650,680,
                    700,720,750,780,800],
    'LoanStatus': [0, 0, 0, 1, 1, 1, 1, 1, 1, 1]}
loan_df = pd.DataFrame(data)

In [20]:
X_loan = loan_df [["Income","CreditScore"]]
y_loan = loan_df ["LoanStatus"]
#create model
model_l = DecisionTreeClassifier(max_depth=3, random_state=42)
#train model
model_l.fit(X_loan,y_loan)
#prediction
p1 = model_l.predict([[25000,420]])
print("P1:", p1)


P1: [0]


##### 3. disease dataset

In [21]:
#disease dataset
data = {"Age":[25,30,35,40,45,50,55,60,42,32],
    "BMI":[21,22,24,27,29,31,33,35,28,19],
    "BP":[110,115,120,130,135,140,145,150,132,118],
    "Disease":[0,0,0,1,1,1,1,1,1,0]}
disease_df = pd.DataFrame(data)

In [22]:
X_disease = disease_df[["Age","BMI","BP"]]
y_disease = disease_df["Disease"]
#create model
model_d = DecisionTreeClassifier(max_depth=3, random_state=42)
#train model
model_d.fit(X_disease,y_disease)
#prediction
p_1 = model_d.predict([[48,26,140]])
print("p1:",p_1)


p1: [1]


### Task 5: Tree Analysis 
1. Identify root node 
2. Identify leaf nodes 
3. Explain decision path 

##### 1. student dataset

In [ ]:
tree_rules_1 = export_text(model_s, feature_names=list(X_student_df.columns))
print(tree_rules_1)

|--- Attendance <= 73.50
|   |--- class: 0
|--- Attendance >  73.50
|   |--- class: 1



- root node  - Attendance.
- Attendance ≤ 73.50 → Fail (0)
- Attendance > 73.50 → Pass (1)
- 2 leaf nodes.
  | Leaf Node | Prediction        |
  | --------- | ----------------- |
  | class: 0  | Result = 0 (Fail) |
  | class: 1  | Result = 1 (Pass) |
- This indicates that attendance is the most influential factor in predicting student results in this dataset

2. ##### loan prediction dataset

In [24]:
tree_rules_2 = export_text(model_l, feature_names=list(X_loan.columns))
print(tree_rules_2)

|--- Income <= 40000.00
|   |--- class: 0
|--- Income >  40000.00
|   |--- class: 1



- Root node – Income
- Income ≤ 40000 → Loan Rejected (0)
- Income > 40000 → Loan Approved (1)
- 2 leaf nodes.

| Leaf Node | Prediction |
|-----------|------------|
| class: 0 | LoanStatus = 0 (Rejected) |
| class: 1 | LoanStatus = 1 (Approved) |

- This indicates that Income is the most influential factor in predicting loan approval in this dataset.

##### 5. disease dataset

In [25]:
tree_rules_3 = export_text(model_d, feature_names=list(X_disease.columns))
print(tree_rules_3)

|--- BMI <= 25.50
|   |--- class: 0
|--- BMI >  25.50
|   |--- class: 1



- Root node – BMI
- BMI ≤ 25.50 → No Disease (0)
- BMI > 25.50 → Disease Present (1)
- 2 leaf nodes.

| Leaf Node | Prediction |
|-----------|------------|
| class: 0 | Disease = 0 (No Disease) |
| class: 1 | Disease = 1 (Disease Present) |

- This indicates that BMI is the most influential factor in predicting disease status in this dataset.

### Task 6: Compare Algorithms 
Compare: 
1. Logistic Regression 
2. KNN 
3. Decision Tree 

 

| Feature                     | Logistic Regression                                       | KNN (K-Nearest Neighbors)                                      | Decision Tree                                           |
| --------------------------- | --------------------------------------------------------- | -------------------------------------------------------------- | ------------------------------------------------------- |
| Algorithm Type              | Linear Classification                                     | Instance-Based / Distance-Based Classification                 | Tree-Based Classification                               |
| Working Principle           | Uses a mathematical model to estimate class probabilities | Classifies based on the majority class among nearest neighbors | Creates decision rules through recursive feature splits |
| Handles Non-Linear Data     | No (unless features are transformed)                      | Yes                                                            | Yes                                                     |
| Feature Scaling Required    | Yes                                                       | Yes (very important)                                           | No                                                      |
| Easy to Interpret           | Moderate                                                  | Difficult                                                      | Very Easy                                               |
| Training Speed              | Fast                                                      | Very Fast (almost no training)                                 | Moderate                                                |
| Prediction Speed            | Fast                                                      | Slow (must compute distances to training samples)              | Fast                                                    |
| Overfitting Risk            | Low to Medium                                             | Depends on K (small K → high risk)                             | High (without pruning)                                  |
| Suitable for Small Datasets | Yes                                                       | Yes                                                            | Yes                                                     |
| Suitable for Large Datasets | Yes                                                       | Poor (slow prediction)                                         | Yes                                                     |
| Visualization               | Difficult                                                 | Difficult                                                      | Easy to Visualize                                       |

